compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/ROCDLPrefetchSharedMemoryCopy.cpp

```diff
@@ -301,105 +234,292 @@ class LoopPrefetcher {
+    opToCluster[op] = clusterID;
   }
+  ++clusterID;
 
-  /// Creates all compute stage ops for a loop iteration with |rewriter| and
-  /// maps the original loop induction variable to |iv| in |mapping|.
-  SmallVector<Value> emitCompute(IRMapping &mapping, RewriterBase &rewriter,
-                                 Value indVar) {
-    // Map the original loop induction variable to |iv| for later op rewrites.
-    mapping.map(forOp.getInductionVar(), indVar);
-
-    SmallVector<Value> results;
-    for (Operation *op : computeStage) {
-      if (auto yieldOp = dyn_cast<scf::YieldOp>(op)) {
-        // On yield, return the operands of the yield converted.
-        results =
-            llvm::map_to_vector<4>(yieldOp.getOperands(), [&](Value operand) {
-              return mapping.lookup(operand);
-            });
-        break;
-      }
+  // Assign cluster ID 2 to all write stage operations
+  for (Operation *op : prefetcher.writeStage) {
+    opToCluster[op] = clusterID;
+  }
 
-      rewriter.clone(*op, mapping);
+  LDBG() << "Built opToCluster map with " << clusterID + 1 << " clusters";
+}
+
+// Builds the final schedule using opToCluster and opToStage mappings.
+// Sorts operations by cluster ID, maintaining original order within each
+// cluster.
+static void buildFinalSchedule(
+    LoopPrefetcher &prefetcher,
+    const llvm::DenseMap<Operation *, unsigned> &opToStage,
+    const llvm::DenseMap<Operation *, unsigned> &opToCluster,
+    std::vector<std::pair<Operation *, unsigned>> &finalSchedule) {
+
+  // Collect all operations from all stages with their cluster IDs
+  SmallVector<Operation *> allOps;
+  allOps.append(prefetcher.readStage.begin(), prefetcher.readStage.end());
+  allOps.append(prefetcher.computeStage.begin(), prefetcher.computeStage.end());
+  allOps.append(prefetcher.writeStage.begin(), prefetcher.writeStage.end());
+
+  // Sort by cluster ID, maintaining original order within each cluster
+  llvm::stable_sort(allOps, [&](Operation *a, Operation *b) {
+    unsigned clusterA = opToCluster.lookup(a);
+    unsigned clusterB = opToCluster.lookup(b);
+    return clusterA < clusterB;
+  });
+
+  // Build the final schedule from the sorted operations
+  for (Operation *op : allOps) {
+    if (opToStage.count(op)) {
+      unsigned stage = opToStage.lookup(op);
+      finalSchedule.push_back({op, stage});
     }
+  }
+
+  LDBG() << "Built final schedule with " << finalSchedule.size()
+         << " operations";
+}
 
-    return results;
+// Builds the PipeliningOption with getScheduleFn.
+static scf::PipeliningOption buildPipeliningOption(
+    const std::vector<std::pair<Operation *, unsigned>> &finalSchedule) {
+  scf::PipeliningOption options;
+  options.getScheduleFn =
+      [finalSchedule](scf::ForOp loop,
+                      std::vector<std::pair<Operation *, unsigned>> &outSched) {
+        outSched = finalSchedule;
+      };
+
+  options.peelEpilogue = true;
+  options.supportDynamicLoops = false;
```

Comment:
TODO for dynamic trip count support. We'll probably want to run the in range analysis to restrict to cases where we know K is large.